src/rng.jl

```diff
@@ -0,0 +1,104 @@
+using Random123
+using Random
+using Distributions
+
+import Base.rand
+import Random.seed!
+import Random123: set_counter!
+
+# Default RNG type for when nothing is specified
+_BASE_RNG = Philox2x
+
+"""
+    TracedRNG{R,T}
+
+Wrapped random number generator from Random123 to keep track of random streams during model evaluation
+"""
+mutable struct TracedRNG{T} <:
+               Random.AbstractRNG where {T<:(Random123.AbstractR123{R} where {R})}
+    "Model step counter"
+    count::Int
+    "Inner RNG"
+    rng::T
+    "Array of keys"
+    keys
+end
+
+"""
+    TracedRNG(r::Random123.AbstractR123)
+
+Initialize TracedRNG with r as the inner RNG 
+"""
+function TracedRNG(r::Random123.AbstractR123)
+    set_counter!(r, 0)
+    return TracedRNG(1, r, typeof(r.key)[])
+end
+
+"""
+    TracedRNG()
+
+Create a default TracedRNG
+"""
+function TracedRNG()
+    r = _BASE_RNG()
+    return TracedRNG(r)
+end
+
+# Connect to the Random API
+Random.rng_native_52(rng::TracedRNG{U}) where {U} = Random.rng_native_52(rng.rng)
+Base.rand(rng::TracedRNG{U}, ::Type{T}) where {U,T} = Base.rand(rng.rng, T)
+
+"""
+    split(key::Integer, n::Integer)
+
+Split key into n new keys
+"""
+function split(key::Integer, n::Integer=1) where {T}
+    return map(i -> hash(key, convert(UInt, i)), 1:n)
```

Comment:
A bit simpler, I think:
```suggestion
    return [hash(key, i) for i in UInt(1):UInt(n)]
```